scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala

```diff
@@ -0,0 +1,200 @@
+  *                        resnet-152-XXXX.params and optionally synset.txt).
+  *                        Supports model loading from various sources like local disk,
+  *                        hdfs, https and s3. file://, hdfs://, https://, s3://
+  * @param inputDescriptors Descriptors defining the input node names, shape,
+  *                         layout and Type parameters
+  * @param outputDescriptors Descriptors defining the output node names, shape,
+  *                          layout and Type parameters
+  */
+class Predictor(modelPathPrefix: String,
+             protected val inputDescriptors: IndexedSeq[DataDesc],
+             protected var outputDescriptors:
+             Option[IndexedSeq[DataDesc]] = None) extends PredictBase {
+
+  private val logger = LoggerFactory.getLogger(classOf[Predictor])
+
+  protected var batchIndex = inputDescriptors(0).layout.indexOf('N')
+  protected var batchSize = if (batchIndex != -1 ) inputDescriptors(0).shape(batchIndex) else 1
+
+  protected var iDescriptors = inputDescriptors
+
+  inputDescriptors.foreach((f: DataDesc) => require(f.layout.indexOf('N') == batchIndex,
+    "batch size should be in the same index for all inputs"))
+
+
+  if (batchIndex != -1) {
+    inputDescriptors.foreach((f: DataDesc) => require(f.shape(batchIndex) == batchSize,
+      "batch size should be same for all inputs"))
+  } else {
+    // TODO: this is assuming that the input needs a batch
+    iDescriptors = inputDescriptors.map((f : DataDesc) => new DataDesc(f.name,
+    Shape(1 +: f.shape.toVector), f.dtype, 'N' +: f.layout) )
+    batchIndex = 1
+  }
+
+  protected val mxNetHandler = MXNetHandler()
+
+  protected val mod = loadModule()
+
+  /**
+    * This method will take input as IndexedSeq one dimensional arrays and creates
+    * NDArray needed for inference. The array will be reshaped based on the input descriptors.
+    *
+    * @param input : A IndexedSequence of Java one-dimensional array, An IndexedSequence is
+    *              is needed when the model has more than one input/output
+    * @return IndexedSequence array of outputs.
+    */
+  override def predict(input: IndexedSeq[Array[Float]]): IndexedSeq[Array[Float]] = {
+
+    require(input.length == inputDescriptors.length, "number of inputs provided: %d" +
+      " do not match number of inputs in inputDescriptors: %d".format(input.length,
+        inputDescriptors.length))
+
+    for((i, d) <- input.zip(inputDescriptors)) {
+      require (i.length == d.shape.product/batchSize, "number of elements:" +
+        " %d in the input does not match the shape:%s".format( i.length, d.shape.toString()))
+    }
+
+    var inputND: ListBuffer[NDArray] = ListBuffer.empty[NDArray]
+
+    for((i, d) <- input.zip(inputDescriptors)) {
+      val shape = d.shape.toVector.patch(from = batchIndex, patch = Vector(1), replaced = 1)
+
+      inputND += mxNetHandler.execute(NDArray.array(i, Shape(shape)))
```

Comment:
I am a bit confused here, `NDArray.array(i, Shape(shape))` is executed by a new thread? 

and it's actually blocking here until the function is finished? why we need to do this?

Reply:
Calls to MXNet's native library are not safe to call from more than one thread per program execution (unless that has been changed). This is just how MXNet works at present. We [described this on the discussion site at one point](https://discuss.mxnet.io/t/fixing-thread-safety-issues-in-scala-library/236).

It's actually not entirely known if some calls are safe and some are not but we've found the best stability from single-threading all calls to MXNet.

Reply:
I am going by Calum's findings here